index.bs

```
@@ -4648,71 +4659,152 @@ The <dfn export for=commands>browsingContext.setViewport</dfn> command modifies
+   viewport=] to be the <code>width</code> of |viewport| in CSS pixels and
+   set the height of the |navigable|'s [=layout viewport=] to be the
+   <code>height</code> of |viewport| in CSS pixels.
+
+1. Otherwise, set the |navigable|'s [=layout viewport=] to the
+   implementation-defined default.
+
+</div>
+
+<div algorithm="set viewport when a top-level traversable is created">
+
+After creating a document in a new [=navigable/top-level traversable=] |navigable| and
+before the [=run WebDriver BiDi preload scripts=] algorithm is invoked:
+
+1. Let |user context| be |navigable|'s [=associated user context=].
+
+1. If [=viewport overrides map=] [=map/contains=] |user context|:
+
+   1. If |viewport overrides map|[|user context|] [=map/contains=] the <code>viewport</code> field:
 
-1. If the implementation is unable to adjust the |navigable|'s [=layout viewport=]
+      1. [=Set viewport=] with |navigable| and |viewport overrides map|[|user context|]["<code>viewport</code>"].
+
+   1. If |viewport overrides map|[|user context|] [=map/contains=] the <code>devicePixelRatio</code> field:
+
+      1. For the |navigable| and all [=descendant navigables=]:
+
+         1. [=Set device pixel ratio override=] with |navigable| and |viewport overrides map|[|user context|]["<code>devicePixelRatio</code>""].
+
+</div>
+
+<div algorithm="remote end steps for browsingContext.setViewport">
+
+The [=remote end steps=] with |command parameters| are:
+
+1. If the implementation is unable to adjust the [=layout viewport=]
    parameters with the given |command parameters| for any reason, return
    [=error=] with [=error code=] [=unsupported operation=].
 
-1. If |command parameters| [=map/contains=] the <code>viewport</code> field:
+1. If |command parameters| [=map/contains=] "<code>userContexts</code>"
+   and |command parameters| [=map/contains=] "<code>context</code>",
+   return [=error=] with [=error code=] [=invalid argument=].
```

We should probably also raise an error if both parameters are missing? It will effectively be a noop and won't crash or anything, but the user should know about it. Otherwise they might assume it just applied setViewport to all navigables?

---

Yeah, that's true, added ✅